leetcode/0001/TwoSum.cs

```diff
@@ -0,0 +1,50 @@
+using System.Collections.Generic;
+
+public class Solution
+{
+    public int[] TwoSum(int[] nums, int target)
+    {
+        return SolveWithOnePass(nums, target);
+    }
+
+    public int[] SolveWithOnePass(int[] nums, int target)
+    {
+        var set = new Dictionary<int, int>(nums.length);
+
+        for(int x = 0; x < nums.Length; x++)
+        {
+            if (set.TryGetValue(nums[x], out var result))
+            {
+                if (x == result) continue;
+
+                return new int[] { x, result };
+            }
+            
+            set[target-nums[x]] = x;
+        }
+
+        return null;
+    }
+
+    public int[] SolveWithDictionary(int[] nums, int target)
+    {
+        var set = new Dictionary<int, int>();
+
+        for (int x = 0; x < nums.Length; x++)
+        {
+            set[target-nums[x]] = x;
```

Comment:
Load up the dictionary with the difference of each number and the target.  This way we can iterate over the list of number and look for the second component of the sum.  The dictionary gives us a fast way to search through the list.  Near-constant time on average.